crates/hotshot/types/src/epoch_membership.rs

```diff
@@ -155,136 +159,237 @@ where
     }
 
     /// Catches the membership up to the epoch passed as an argument.  
-    /// To do this try to get the stake table for the epoch containing this epoch's root
-    /// if the root does not exist recursively catchup until you've found it
+    /// To do this, try to get the stake table for the epoch containing this epoch's root and
+    /// the stake table for the epoch containing this epoch's drb result.
+    /// If they do not exist, then go one by one back until we find a stake table.
     ///
     /// If there is another catchup in progress this will not duplicate efforts
-    /// e.g. if we start with only epoch 0 stake table and call catchup for epoch 10, then call catchup for epoch 20
+    /// e.g. if we start with only the first epoch stake table and call catchup for epoch 10, then call catchup for epoch 20
     /// the first caller will actually do the work for to catchup to epoch 10 then the second caller will continue
     /// catching up to epoch 20
-    async fn catchup(self, epoch: TYPES::Epoch) -> Result<EpochMembership<TYPES>> {
-        // recursively catchup until we have a stake table for the epoch containing our root
-        ensure!(
-            *epoch != 0 && *epoch != 1,
-            "We are trying to catchup to epoch 0! This means the initial stake table is missing!"
-        );
-        let root_epoch = TYPES::Epoch::new(*epoch - 2);
-
-        let root_membership = if self.membership.read().await.has_stake_table(root_epoch) {
-            EpochMembership {
-                epoch: Some(root_epoch),
-                coordinator: self.clone(),
-            }
-        } else {
-            Box::pin(self.wait_for_catchup(root_epoch)).await?
+    async fn catchup(self, epoch: TYPES::Epoch, epoch_tx: Sender<Result<EpochMembership<TYPES>>>) {
```

Comment:
This function has become really long and it's hard to follow.  I think we should split it as best we can into parts.

Logically this function does the following:

1. Find all the epochs we need to fetch a stake table for
2. Try to fetch the stake table, or wait for another task to, for all those epochs
3. Try to catch up DRB
4. clean up the map, sending any error encountered to the channels we have the sender for

I think we should at least split some of these parts out into separate functions.